File: server/services/documentPdfCache.js
// 统一官方文档 PDF 持久化缓存：IPO 招股书、上市公告书及其他公告共用。
// 缓存只保存可重新获取的官方原文，不承载业务事实；业务事实仍写入 raw_records/event.documents。
const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

const DEFAULT_TTL_DAYS = 30;
const DEFAULT_MAX_BYTES = 5 * 1024 * 1024 * 1024;

function cacheDir() {
  return process.env.DOCUMENT_PDF_CACHE_DIR
    || process.env.ARBITRAGE_PDF_CACHE_DIR
    || path.join(__dirname, '..', '..', 'data', 'document_pdf_cache');
}

function cacheLimits() {
  const ttlValue = process.env.DOCUMENT_PDF_CACHE_TTL_DAYS ?? process.env.ARBITRAGE_PDF_CACHE_TTL_DAYS;
  const maxValue = process.env.DOCUMENT_PDF_CACHE_MAX_BYTES ?? process.env.ARBITRAGE_PDF_CACHE_MAX_BYTES;
  const ttl = Number(ttlValue);
  const max = Number(maxValue);
  return {
    ttlDays: Number.isFinite(ttl) && ttl > 0 ? Math.floor(ttl) : DEFAULT_TTL_DAYS,
    maxBytes: Number.isFinite(max) && max >= 1024 * 1024 ? Math.floor(max) : DEFAULT_MAX_BYTES,
  };
}

function cachePath(url, directory = cacheDir()) {
  const digest = crypto.createHash('sha256').update(String(url || '')).digest('hex');
  return path.join(directory, `${digest}.pdf`);
}

function cleanupDocumentPdfCache(now = Date.now(), directoryOverride = null) {
  const directory = directoryOverride || cacheDir();
  fs.mkdirSync(directory, { recursive: true });
  const { ttlDays, maxBytes } = cacheLimits();
  const cutoff = now - ttlDays * 24 * 60 * 60 * 1000;
  const files = [];
  let deletedExpired = 0;
  let deletedOverflow = 0;
  for (const name of fs.readdirSync(directory)) {
    const file = path.join(directory, name);
    let stat;
    try { stat = fs.statSync(file); } catch (_) { continue; }
    const ext = path.extname(name).toLowerCase();
    if (!stat.isFile() || !['.pdf', '.part'].includes(ext)) continue;
    if (stat.mtimeMs < cutoff) {
      try { fs.unlinkSync(file); deletedExpired++; } catch (_) {}
      continue;
    }
    if (ext === '.pdf') files.push({ file, mtimeMs: stat.mtimeMs, size: stat.size });
  }
  let totalBytes = files.reduce((sum, item) => sum + item.size, 0);
  files.sort((a, b) => a.mtimeMs - b.mtimeMs || a.file.localeCompare(b.file));
  for (const item of files) {
    if (totalBytes <= maxBytes) break;
    try { fs.unlinkSync(item.file); totalBytes -= item.size; deletedOverflow++; } catch (_) {}
  }
  return {
    directory, ttlDays, maxBytes, totalBytes,
    fileCount: Math.max(files.length - deletedOverflow, 0), deletedExpired, deletedOverflow,
  };
}

function readCachedPdf(url, { now = Date.now(), directory = cacheDir() } = {}) {
  cleanupDocumentPdfCache(now, directory);
  const file = cachePath(url, directory);
  try {
    const stat = fs.statSync(file);
    const { ttlDays } = cacheLimits();
    if (!stat.isFile() || stat.size <= 0 || stat.mtimeMs < now - ttlDays * 24 * 60 * 60 * 1000) return null;
    const buffer = fs.readFileSync(file);
    return buffer.length ? buffer : null;
  } catch (_) {
    return null;
  }
}

function writeCachedPdf(url, buffer, { directory = cacheDir() } = {}) {
  if (!Buffer.isBuffer(buffer) || !buffer.length) throw new Error('PDF 缓存内容为空');
  fs.mkdirSync(directory, { recursive: true });
  const target = cachePath(url, directory);
  const temp = `${target}.${process.pid}.${Date.now()}.part`;
  try {
    fs.writeFileSync(temp, buffer, { mode: 0o600 });
    fs.renameSync(temp, target);
  } finally {
    try { if (fs.existsSync(temp)) fs.unlinkSync(temp); } catch (_) {}
  }
  cleanupDocumentPdfCache(Date.now(), directory);
  return target;
}

module.exports = {
  DEFAULT_TTL_DAYS,
  DEFAULT_MAX_BYTES,
  cacheDir,
  cacheLimits,
  cachePath,
  cleanupDocumentPdfCache,
  readCachedPdf,
  writeCachedPdf,
};

File: ipo-report/document_pdf_cache.py
"""Node/Python 共用的官方文档 PDF 持久化缓存。"""
import hashlib
import os
import tempfile
import time
from pathlib import Path

DEFAULT_TTL_DAYS = 30
DEFAULT_MAX_BYTES = 5 * 1024 * 1024 * 1024


def cache_dir(cache_dir=None):
    configured = str(cache_dir or os.environ.get("DOCUMENT_PDF_CACHE_DIR")
                     or os.environ.get("ARBITRAGE_PDF_CACHE_DIR") or "").strip()
    if configured:
        return Path(configured)
    return Path(__file__).resolve().parents[1] / "data" / "document_pdf_cache"


def cache_limits():
    try:
        ttl_days = int(os.environ.get("DOCUMENT_PDF_CACHE_TTL_DAYS",
                                      os.environ.get("ARBITRAGE_PDF_CACHE_TTL_DAYS", DEFAULT_TTL_DAYS)))
    except (TypeError, ValueError):
        ttl_days = DEFAULT_TTL_DAYS
    try:
        max_bytes = int(os.environ.get("DOCUMENT_PDF_CACHE_MAX_BYTES",
                                      os.environ.get("ARBITRAGE_PDF_CACHE_MAX_BYTES", DEFAULT_MAX_BYTES)))
    except (TypeError, ValueError):
        max_bytes = DEFAULT_MAX_BYTES
    return max(ttl_days, 1), max(max_bytes, 1 * 1024 * 1024)


def pdf_cache_path(url, directory=None):
    digest = hashlib.sha256(str(url or "").encode("utf-8")).hexdigest()
    return cache_dir(directory) / f"{digest}.pdf"


def cleanup_pdf_cache(directory=None, now=None):
    directory = cache_dir(directory)
    directory.mkdir(parents=True, exist_ok=True)
    ttl_days, max_bytes = cache_limits()
    current = float(time.time() if now is None else now)
    cutoff = current - ttl_days * 24 * 60 * 60
    files = []
    deleted_expired = 0
    deleted_overflow = 0
    for path in directory.glob("*"):
        if not path.is_file() or path.suffix.lower() not in {".pdf", ".part"}:
            continue
        try:
            stat = path.stat()
        except OSError:
            continue
        if stat.st_mtime < cutoff:
            try:
                path.unlink()
                deleted_expired += 1
            except OSError:
                pass
            continue
        if path.suffix.lower() == ".pdf":
            files.append((stat.st_mtime, stat.st_size, path))
    total_bytes = sum(size for _, size, _ in files)
    for _, size, path in sorted(files, key=lambda item: (item[0], str(item[2]))):
        if total_bytes <= max_bytes:
            break
        try:
            path.unlink()
            total_bytes -= size
            deleted_overflow += 1
        except OSError:
            pass
    return {
        "directory": str(directory), "total_bytes": max(total_bytes, 0),
        "file_count": max(len(files) - deleted_overflow, 0),
        "deleted_expired": deleted_expired, "deleted_overflow": deleted_overflow,
        "ttl_days": ttl_days, "max_bytes": max_bytes,
    }


def get_cached_pdf(url, directory=None, now=None):
    directory = cache_dir(directory)
    cleanup_pdf_cache(directory, now=now)
    path = pdf_cache_path(url, directory)
    try:
        stat = path.stat()
    except OSError:
        return None
    ttl_days, _ = cache_limits()
    current = float(time.time() if now is None else now)
    if stat.st_size <= 0 or stat.st_mtime < current - ttl_days * 24 * 60 * 60:
        return None
    return path


def put_cached_pdf(url, content, directory=None):
    if not content:
        raise ValueError("PDF 缓存内容为空")
    directory = cache_dir(directory)
    directory.mkdir(parents=True, exist_ok=True)
    target = pdf_cache_path(url, directory)
    fd, temp_path = tempfile.mkstemp(prefix="document-", suffix=".part", dir=str(directory))
    try:
        with os.fdopen(fd, "wb") as handle:
            handle.write(content)
        os.replace(temp_path, target)
        try:
            os.chmod(target, 0o600)
        except OSError:
            pass
    finally:
        try:
            if os.path.exists(temp_path):
                os.remove(temp_path)
        except OSError:
            pass
    cleanup_pdf_cache(directory)
    return target


__all__ = ["DEFAULT_TTL_DAYS", "DEFAULT_MAX_BYTES", "cache_dir", "cache_limits",
           "pdf_cache_path", "cleanup_pdf_cache", "get_cached_pdf", "put_cached_pdf"]
